mirdata/core.py

```diff
@@ -0,0 +1,232 @@
+
+        if data_home is None:
+            self.data_home = self.default_path
+        else:
+            self.data_home = data_home
+
+        # this is a hack to be able to have dataset-specific docstrings
+        self.track = lambda track_id: self._track(track_id)
+        self.track.__doc__ = self._track_object.__doc__  # set the docstring
+
+        # inherit any public load functions from the module
+        for method_name in dir(module):
+            if method_name.startswith("load_"):
+                method = getattr(module, method_name)
+                setattr(self, method_name, method)
+                # getattr(self, method_name).__doc__ = method.__doc__
+
+    def __repr__(self):
+        repr_string = "The {} dataset\n".format(self.name)
+        repr_string += "-" * MAX_STR_LEN
+        repr_string += "\n"
+        repr_string += (
+            "Call the .readme method for complete documentation of this dataset.\n"
+        )
+        repr_string += "Call the .cite method for bibtex citations.\n"
+        repr_string += "-" * MAX_STR_LEN
+        repr_string += "\n"
+        if self._track_object is not None:
+            repr_string += self.track.__doc__
+            repr_string += "-" * MAX_STR_LEN
+            repr_string += "\n"
+
+        return repr_string
+
+    @property
+    def default_path(self):
+        """Get the default path for the dataset
+
+        Returns:
+            default_path (str): Local path to the dataset
+        """
+        mir_datasets_dir = os.path.join(os.getenv("HOME", "/tmp"), "mir_datasets")
+        return os.path.join(mir_datasets_dir, self.name)
+
+    def _track(self, track_id):
+        """Load a track by track_id.
+        Hidden helper function that gets called as a lambda.
+
+        Args:
+            track_id (str): track id of the track
+        
+        Returns:
+            track (dataset.Track): an instance of this dataset's Track object
+        """
+        if self._track_object is None:
+            raise NotImplementedError
+        else:
+            return self._track_object(track_id, self.data_home)
+
+    def load_tracks(self):
+        """Load all tracks in the dataset
+
+        Returns:
+            (dict): {`track_id`: track data}
+        
+        Raises:
+            NotImplementedError: If the dataset does not support Track objects
+        """
+        return {track_id: self.track(track_id) for track_id in self.track_ids}
+
+    def choice_track(self):
+        """Choose a random track
+
+        Returns:
+            track (dataset.Track): a random Track object
+        """
+        return self.track(random.choice(self.track_ids))
+
+    def readme(self):
+        """Print the dataset's readme.
+        """
+        print(self._readme_str)
+
+    def cite(self):
+        """Print the reference"""
+        print("========== BibTeX ==========")
+        print(self.bibtex)
+
+    def download(self, partial_download=None, force_overwrite=False, cleanup=True):
```

Comment:
does this class overwrite the download inside the loader? 
in Saraga we need to adapt the download to query the multi-track audio files individually.

Reply:
for loaders that need a non-standard download method, you can set a `_download` method in the module and it uses that instead. In this PR see for example maestro.py